File: python/Binary Tree Zigzag Level Order Traversal/Binary_Tree_Zigzag_Level_Order_Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        if not root :
            return []
        left_to_right = True
        res =[]
        level = []
        queue = []
        queue.append(root)
        queue.append(None)
        while queue:
            root = queue.pop(0)
            if root is None:
                if queue:
                    queue.append(None)
                res.append(level)
                level = []
                left_to_right = not left_to_right
            else:
                if left_to_right:
                    level.append(root.val)
                else:
                    level.insert(0,root.val)
                if root.left is not None:
                    queue.append(root.left)
                if root.right is not None:
                    queue.append(root.right)
        return res

File: leetcode/cpp/BinaryTreeZigzagLevelOrderTraversal.cpp
//
// Created by zhangqi on 2018/1/31.

// Time:  O(n)
// Space: O(n)

#include <vector>
#include <queue>
#include <algorithm>
#include "TreeNode.h"

using namespace std;


class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        queue<TreeNode*> que;
        int leftToRight = 0;
        if(root){
            que.emplace(root);
        }
        while (!que.empty()){
            vector<int> level;
            int size = que.size();
            for(int i = 0 ; i < size; i++){
                TreeNode *front = que.front();
                que.pop();
                level.push_back(front->val);
                if(front->left){
                    que.emplace(front->left);
                }
                if(front->right){
                    que.emplace(front->right);
                }
            }
            if(leftToRight++ % 2 == 1){
                reverse(level.begin(), level.end());
            }
            result.push_back(level);
        }
        return result;
    }
};